### PR TITLE
Android: reconcile UI text with editor text

### PR DESCRIPTION
**Pull request by jmartinesp**
This should help in cases where the UI text is modified externally by the OS or some app.

Basically, what we do here is we intercept any change made outside of the editor by either the OS or the implementing app, pass them to the editor, apply them, then notify the text watchers of the result of these editor changes.

It seems to work well so far, the tests pass. We probably need to thoroughly test it with some day-to-day usage.

PS: this is heavily based on @ganfra 's work [here](https://github.com/matrix-org/matrix-rich-text-editor/pull/940). Thanks!

## Video example

https://github.com/matrix-org/matrix-rich-text-editor/assets/480955/33234304-038f-40c5-bbd3-b00bd3ffc7f0

### CODE REVIEW

File: platforms/android/library/src/main/java/io/element/android/wysiwyg/EditorEditText.kt [hunk 1]
```diff
@@ -615,3 +637,61 @@ private fun KeyEvent.isMovementKey(): Boolean {
...
+        nestedWatchers.forEach { it.beforeTextChanged(s, start, count, after) }
+    }
+
+    fun notifyOnTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
+        nestedWatchers.forEach { it.onTextChanged(s, start, before, count) }
+    }
+
+    fun notifyAfterTextChanged(s: Editable?) {
+        nestedWatchers.forEach { it.afterTextChanged(s) }
+    }
```
Comment (jmartinesp) [inline platforms/android/library/src/main/java/io/element/android/wysiwyg/EditorEditText.kt:696]: I don't really like this solution, to be honest. I'd rather use the original overridden methods, but I couldn't find a good way to do it.

File: platforms/android/library/src/main/java/io/element/android/wysiwyg/EditorEditText.kt [hunk 2]
```diff
@@ -578,6 +582,18 @@ class EditorEditText : AppCompatEditText {
         )
     }
 
+    override fun addTextChangedListener(watcher: TextWatcher) {
+        if (!this::textWatcher.isInitialized) {
```
Comment (jmartinesp) [inline platforms/android/library/src/main/java/io/element/android/wysiwyg/EditorEditText.kt:586]: Thanks @ganfra for this, I didn't really remember how to handle code that was called when a view is inflated and was stuck with a null pointer exception.

File: platforms/android/library/src/main/java/io/element/android/wysiwyg/EditorEditText.kt [hunk 3]
```diff
@@ -578,6 +581,18 @@ class EditorEditText : AppCompatEditText {
...
+            this.textWatcher = EditorTextWatcher()
+            super.addTextChangedListener(this.textWatcher)
+        }
+        textWatcher.addChild(watcher)
+    }
+
+    override fun removeTextChangedListener(watcher: TextWatcher) {
+        textWatcher.removeChild(watcher)
```
Comment (ganfra) [inline platforms/android/library/src/main/java/io/element/android/wysiwyg/EditorEditText.kt:593]: If there is no child, remove the textWatcher from parent too?

File: platforms/android/library/src/main/java/io/element/android/wysiwyg/EditorTextWatcher.kt [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+import android.text.Editable
+import android.text.TextWatcher
+import timber.log.Timber
+import java.util.concurrent.atomic.AtomicBoolean
+
+/**
+ * A [TextWatcher] that intercepts changes in the underlying text and can have child watchers.
+ */
+class EditorTextWatcher: TextWatcher {
```
Comment (ganfra) [inline platforms/android/library/src/main/java/io/element/android/wysiwyg/EditorTextWatcher.kt:11]: internal?

**Review by ganfra — APPROVED** [review]
Small questions, otherwise LGTM, happy to see that my previous work is useful :D

File: platforms/android/library/src/main/java/io/element/android/wysiwyg/EditorEditText.kt (hunk 3 above)
Comment (jmartinesp) [reply platforms/android/library/src/main/java/io/element/android/wysiwyg/EditorEditText.kt:593]: I don't think it's worth it, to be honest. Keeping it there has no cost as far as I can tell, and will also be needed to notify the composer of any changes made by any external actors as the spell checker or some suggestion replacements.

**Pull request merged**